addons/sale_timesheet_margin/models/sale_order_line.py

```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+from odoo import api, models
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    @api.depends('analytic_line_ids.amount')
+    def _compute_purchase_price(self):
+        timesheet_sol = self.filtered(lambda sol: sol.qty_delivered_method == 'timesheet')
+        super(SaleOrderLine, self - timesheet_sol)._compute_purchase_price()
+        if timesheet_sol:
+            group_amount = self.env['account.analytic.line'].read_group(
+                [('so_line', 'in', timesheet_sol.ids), ('project_id', '!=', False)],
+                ['so_line', 'amount:sum', 'unit_amount:sum'],
+                ['so_line'])
+            mapped_sol_timesheet_amount = {
+                amount['so_line'][0]: -amount['amount'] / amount['unit_amount']
+                for amount in group_amount
+            }
+            for line in timesheet_sol:
+                if not line.product_id:
+                    line.purchase_price = 0.0
+                    continue
```

Comment:
Unnecessary since line have sol.qty_delivered_method == 'timesheet' and thus, product_id should be set.

Reply:
Yes, ok I understood. I'm agree with you, this condition is unnecessary 🙂